README.md
[![Community Extension](https://img.shields.io/badge/Community%20Extension-An%20open%20source%20community%20maintained%20project-FF4700)](https://github.com/camunda-community-hub/community)
[![Lifecycle: Incubating](https://img.shields.io/badge/Lifecycle-Incubating-blue)](https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md#incubating-)
![Compatible with: Camunda Platform 7](https://img.shields.io/badge/Compatible%20with-Camunda%20Platform%207-26d07c)
![Compatible with: Camunda Platform 8](https://img.shields.io/badge/Compatible%20with-Camunda%20Platform%208-0072Ce)
![Educational Tooling](https://img.shields.io/badge/Educational%20Tooling-Project%20for%20getting%20started%20with%20Camunda%20for%20educators-%239F2B68)
![Maven Central](https://img.shields.io/nexus/r/https/s01.oss.sonatype.org/io.github.pme123/camunda-dmn-tester-shared_2.13.svg)
![GitHub repo size](https://img.shields.io/github/repo-size/pme123/camunda-dmn-tester)
![Docker Pulls](https://img.shields.io/docker/pulls/pame/camunda-dmn-tester)

# Camunda DMN Table Tester
A little DMN Table tester with the following Goals:
* As a developer I want to test the DMNs that I get from the Business, even not knowing the concrete rules.
* Business people can create their own tests.
* They can easily adjust the tests to the dynamic nature of DMN Tables.

> This is exactly the kind of tool that as soon as I saw it, I knew it was something I was missing.
>
> I really enjoy being able to help people especially new people with modeling DMN, so I often have people sending me models to validate.
> 
> This tool’s fundamental helpfulness comes first from being able to quickly catch some of the simple common errors, but also in the way it makes it easy for business users to explore the wide variety of possible ways in DMN that decisions can be modeled. This was in fact the original goal of DMN and this is a strong step towards narrowing the Business IT Gap for the standard.
>  
> **Niall Deehan** Developer Relations Camunda

## Usage
I wrote two blog article that explains how you can use it:

* [Testing (Camunda)-DMN Tables automatically](https://pme123.medium.com/testing-camunda-dmn-tables-automatically-713497ab57e6)

* [Use the DMN Tester for Continuous Integration CI](https://pme123.medium.com/testing-camunda-dmn-tables-automatically-part-2-d3931ed38f51)

And there is a Recording from Camunda Community Summit:
* [Testing DMN Tables Automatically](https://page.camunda.com/ccs-mengelt-testing-dmn-tables-automatically) 

## Test Configuration
![Config Dialog](docs/Screenshot_configDialog.png)
**1 Test Mode**
  
  You can configure a DMN Table Test as:

  - a Unit test (the inputs are not created from evaluating the depending decisions - you define them directly)
  - with all its dependent Inputs (decisions like DMN Tables or Literal Expressions).

**2 Decision Id**

  The id of your table - be aware this is not the name
  - you only find it in the 'Edit DRD' Panel with the DMN Table Object selected.

**3 Path of the DMN**
  
  This path depends on your Docker Image (check _runDmnTester.sh_).
  
  Example: `-v $(pwd)/dmns:/opt/docker/dmns \` 
  - This assumes your DMNs are in a folder _dmns_.
  - So your Path would be `dmns/MyDmn.dmn`.

**4 Test Inputs**

  Add the following for each Input:
  - Key: This is what you need in your input expressions.
  - Type: The Type specified.
  - Values: A comma separated list of all possible inputs.
  - Null Value: If it is possible that the input may be _null_ (not required), check this box.

**5 Test Variables** 

  You can define Variables you use in your Rules. Usually one value is enough.

## Technologies
This projects builds on cool Open Source Projects. So my thanks go to:

### Shared
* [Autowire](https://github.com/lihaoyi/autowire):
  > Autowire is a pair of macros that allows you to perform type-safe, reflection-free RPC between Scala systems.
* [BooPickle](https://boopickle.suzaku.io):
  > BooPickle is the fastest and most size efficient serialization (aka pickling) library that works on both Scala and Scala.js.

### Client
* [Slinky](https://slinky.dev)
  > Write React apps in Scala just like you would in ES6
* [Scalably Typed](https://scalablytyped.org)
  > The Javascript ecosystem for Scala.js!
  I used the facades for Ant Design
* [Ant Design](https://ant.design)
  >A design system for enterprise-level products. Create an efficient and enjoyable work experience.

### Server
* [Scala DMN](https://github.com/camunda/dmn-scala)
  > An engine to execute decisions according to the DMN 1.1 specification.
* [http4s](https://http4s.org)
  > Typeful, functional, streaming HTTP for Scala.
* [ZIO Config](https://zio.github.io/zio-config/)
  > A functional, composable ZIO interface to configuration
* [ZIO](https://zio.dev)
  > Type-safe, composable asynchronous and concurrent programming for Scala

### Start Script
* [Ammonite](https://ammonite.io/#Ammonite)
  > Ammonite lets you use the Scala language for scripting purposes: in the REPL, as scripts, as a library to use in existing projects, or as a standalone systems shell.
## Development
### Server
`sbt server/run`

This starts the Web Server on **Port 8883**.

>This copies the client assets to the classpath of the server.
> So make sure you run `build` before.
>
> Or use the client as described in the next chapter.

### Client
`sbt dev`

This will watch all your changes in the client and automatically refresh your Browser Session.

Open in the Browser **http://localhost:8024**.

## Releasing
Just run `amm ./publish-release.sc VERSION`.

Due to problems with the `"org.xerial.sbt" % "sbt-sonatype"` Plugin you have to release manually:
- https://s01.oss.sonatype.org/#stagingRepositories
  - login
  - check Staging Repository
  - hit _close_ Button
  - hit _release_ Button
    
> if you do not see any of the buttons or repository hit the _refresh_ Button.

### Local publish
At the moment there are 2 steps:
1. Build the Client (full optimization) and the Server:

   `sbt release`
2. Publish: 
   
   `sbt publishLocal`

### Docker
There are 2 Docker Images:

1. The DMN Tester App:

   `sbt server/docker:publishLocal` creates a Docker Image - see also next chapter.

2. The Unit Test Creator:
   See its [README.md](docker/README.md)
   
## Try it out
There are Docker Images you can use with an example in the `demo` directory.

`cd demo`

See the according [README](demo/README.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
* Types of Changes (L3):
  * Added: new features
  * Changed: changes in existing functionality
  * Deprecated: soon-to-be-removed features
  * Removed: now removed features
  * Fixed: any bug fixes
  * Security: in case of vulnerabilities


The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

[](#0-14-0)
## 0.14.0 - 2022-03-13
### Fixed
Fixed problem with dependent Tables to show correct matched rule.
### Added
Support for _date and time("2021-12-12T00:00")_.

[](#0-12-0)
## 0.12.0 - 2021-06-27
### Fixed
Unchecked `nullValue` throw hidden exception in configList.

[](#0-11-0)
## 0.11.0 - 2021-05-22
Added the possibility to use null values as inputs.
[](#0-10-0)
## 0.10.0 - 2021-04-18
Initial Version


project/Settings.scala
import com.typesafe.sbt.packager.Keys._
import com.typesafe.sbt.packager.docker.DockerPlugin
import com.typesafe.sbt.packager.docker.DockerPlugin.autoImport.Docker
import org.portablescala.sbtplatformdeps.PlatformDepsPlugin.autoImport._
import org.scalablytyped.converter.plugin.ScalablyTypedPluginBase.autoImport._
import org.scalajs.sbtplugin.ScalaJSPlugin.autoImport._
import sbt.Keys.{logLevel, _}
import sbt.{Level, _}
import scalajsbundler.sbtplugin.ScalaJSBundlerPlugin.autoImport._

object Settings {

  val projectName = "camunda-dmn-tester"
  lazy val projectVersion = scala.io.Source.fromFile("version").mkString.trim

  lazy val projectSettings: Project => Project =
    _.settings(
      organization := "io.github.pme123",
      version := projectVersion,
      scalaVersion := "2.13.4"
    )

  lazy val ReleaseCmd = Command.command("release") { state =>
    "clean" ::
      "build" ::
      "server/compile" ::
      "server/stage" ::
      state
  }

  lazy val publicationSettings: Project => Project = _.settings(
    publishMavenStyle := true,
    pomIncludeRepository := { _ => false },
    publishTo := {
      val nexus = "https://s01.oss.sonatype.org/"
      if (isSnapshot.value) Some("snapshots" at nexus + "content/repositories/snapshots")
      else Some("releases" at nexus + "service/local/staging/deploy/maven2")
    },
    credentials += Credentials(Path.userHome / ".sbt" / "sonatype_credentials"),
    licenses += ("MIT", url("http://opensource.org/licenses/MIT")),
    homepage := Some(url("https://github.com/pme123/camunda-dmn-tester")),
    startYear := Some(2020),
    logLevel := Level.Debug,
    scmInfo := Some(
      ScmInfo(
        url("https://github.com/pme123/camunda-dmn-tester"),
        "scm:git:github.com:/pme123/camunda-dmn-tester"
      )
    ),
    developers := List(
      Developer(
        id    = "pme123",
        name  = "Pascal Mengelt",
        email = "pascal.mengelt@gmail.com",
        url   = url("https://github.com/pme123")
      )
    )
  )

  lazy val preventPublication: Project => Project =
    _.settings(
      publish := {},
      publishTo := Some(
        Resolver
          .file("Unused transient repository", target.value / "fakepublish")
      ),
      publishArtifact := false,
      publishLocal := {},
      packagedArtifacts := Map.empty
    ) // doesn't work - https://github.com/sbt/sbt-pgp/issues/42

  object shared {
    lazy val deps: Project => Project =
      _.settings(
        libraryDependencies ++= Seq(
          "com.lihaoyi" %%% "autowire" % "0.3.2",
          "io.suzaku" %%% "boopickle" % "1.3.2",
          "com.lihaoyi" %%% "upickle" % "1.4.0",
          "com.lihaoyi" %%% "ujson" % "1.4.0"
        )
      ) //.withoutSuffixFor(JVMPlatform)
  }

  object server {
    lazy val settings: Project => Project = _.settings(
      name := s"$projectName-server",
      Compile / unmanagedResourceDirectories += baseDirectory.value / "../client/target/build",
      testFrameworks += new TestFramework("zio.test.sbt.ZTestFramework"),
      resolvers += Resolver.mavenLocal, // only needed for dmn-engine SNAPSHOT
      //    crossScalaVersions := Deps.supportedScalaVersions
      Test / unmanagedSourceDirectories += baseDirectory.value / "target" / "generated-src"
    )

    lazy val serverDeps: Project => Project =
      _.settings(
        libraryDependencies ++= Seq(
          Deps.http4sDsl,
          Deps.http4sServer,
          Deps.logback
        )
      )

    lazy val deps: Project => Project = _.settings(
      libraryDependencies ++= Seq(
        Deps.ammonite,
        Deps.dmnScala,
        Deps.zio,
        Deps.zioCats,
        Deps.zioConfigHocon,
        Deps.zioConfigMagnolia,
        Deps.zioTest % Test,
        Deps.zioTestSbt % Test,
        Deps.scalaTest % Test
      )
    )

    lazy val docker: Project => Project =
      _.settings(
        dockerBaseImage := "openjdk:11", //eed3si9n/sbt:jdk11-alpine",
        dockerExposedPorts ++= Seq(8883),
        packageName in Docker := projectName,
        dockerUsername := Some("pame"),
        dockerUpdateLatest := true
      ).enablePlugins(DockerPlugin)
  }

  object client {

    lazy val slinkyBasics: Project => Project =
      _.settings(
        scalacOptions += "-Ymacro-annotations",
        requireJsDomEnv in Test := true,
        addCommandAlias(
          "dev",
          "fastOptJS::startWebpackDevServer;~fastOptJS"
        ),
        addCommandAlias("build", "fullOptJS::webpack"),
        libraryDependencies ++= Seq(
          "me.shadaj" %%% "slinky-web" % "0.6.7",
          "me.shadaj" %%% "slinky-hot" % "0.6.7"
        ),
        libraryDependencies ++= Seq(
          "org.scalatest" %%% "scalatest" % "3.1.1" % Test
        ),
        Compile / npmDependencies ++= Seq(
          "react" -> "16.13.1",
          "react-dom" -> "16.13.1",
          "react-proxy" -> "1.1.8"
        ),
        Compile / npmDevDependencies ++= Seq(
          "file-loader" -> "6.0.0",
          "style-loader" -> "1.2.1",
          "css-loader" -> "3.5.3",
          "html-webpack-plugin" -> "4.3.0",
          "copy-webpack-plugin" -> "5.1.1",
          "webpack-merge" -> "4.2.2"
        )
      )

    lazy val antdSettings: Project => Project =
      _.settings(
        stFlavour := Flavour.Slinky,
        useYarn := true,
        stIgnore := List("react-proxy"),
        Compile / npmDependencies ++= Seq(
          "antd" -> "4.5.1", //"4.7.0",
          "@types/react" -> "16.9.42",
          "@types/react-dom" -> "16.9.8"
        )
      )

    lazy val webpackSettings: Project => Project =
      _.settings(
        webpackDevServerPort := 8024,
        version in webpack := "4.43.0",
        version in startWebpackDevServer := "3.11.0",
        webpackResources := baseDirectory.value / "webpack" * "*",
        webpackConfigFile in fastOptJS := Some(
          baseDirectory.value / "webpack" / "webpack-fastopt.config.js"
        ),
        webpackConfigFile in fullOptJS := Some(
          baseDirectory.value / "webpack" / "webpack-opt.config.js"
        ),
        webpackConfigFile in Test := Some(
          baseDirectory.value / "webpack" / "webpack-core.config.js"
        ),
        webpackDevServerExtraArgs in fastOptJS := Seq(
          "--inline",
          "--hot",
          "--disableHostCheck"
        ),
        webpackBundlingMode in fastOptJS := BundlingMode.LibraryOnly()
      )

  }

}
